extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,685 @@
+        .then(adResponse => {
+          // Validate when we have a signature and we have native crypto.
+          if (adResponse.signature && verifySignatureIsAvailable()) {
+            try {
+              // Among other things, the signature might not be proper base64.
+              return verifySignature(adResponse.creativeArrayBuffer,
+                                     base64ToByteArray(adResponse.signature),
+                                     rsaPubKeys);
+            } catch (e) {}
+          }
+          return false;
+        });
+  }
+
+  /**
+   * Render the validated AMP creative directly in the parent page.
+   * @param {boolean} valid If the ad response signature was valid.
+   * @return {Promise<boolean>} Whether the creative was successfully
+   *     rendered.
+   * @private
+   */
+  maybeRenderAmpAd_(valid, bytes) {
+    if (!valid) {
+      return Promise.resolve(false);
+    }
+    // Timer id will be set if we have entered layoutCallback at which point
+    // 3p throttling count was incremented.  We want to "release" the throttle
+    // immediately since we now know we are not a 3p ad.
+    if (this.timerId_) {
+      decrementLoadingAds(this.timerId_, this.getWin());
+    }
+    return xhrFor(this.getWin()).utf8FromArrayBuffer(bytes).then(creative => {
+      // Find the json blob located at the end of the body and parse it.
+      const creativeMetaData = this.getAmpAdMetadata_(creative);
+      if (!creativeMetaData || !this.supportsShadowDom()) {
+        // Shadow DOM is not supported or could not find appropriate markers
+        // within the creative therefore load within cross domain iframe.
+        // Iframe is created immediately (as opposed to waiting for
+        // layoutCallback) as the the creative has been verified as AMP and
+        // will run efficiently.
+        this.renderViaIframe_();
+        this.rendered_ = true;
+        return Promise.resolve(true);
+      } else {
+        try {
+          // Do extraction processing on CSS and body before creating the
+          // shadow root so that if they error out, we don't actually edit
+          // the doc.
+          const cssBlock = this.formatCSSBlock_(creative, creativeMetaData);
+          const bodyBlock = this.formatBody_(creative, creativeMetaData);
+          // Copy fonts to host document head.
+          this.relocateFonts_(creativeMetaData);
+          // Add extensions to head.
+          this.addCreativeExtensions_(creativeMetaData);
+          // Finally, add body and re-formatted CSS styling to the shadow root.
+          const shadowRoot =
+              this.element.shadowRoot || this.element.createShadowRoot();
+          shadowRoot./*REVIEW*/innerHTML += (cssBlock + bodyBlock);
+          this.rendered_ = true;
+          this.onAmpCreativeShadowDomRender();
+          return Promise.resolve(true);
+        } catch (e) {
+          // If we fail on any of the steps of Shadow DOM construction, just
```

Comment:
Okay -- one of our team members is working on a cleanup of a number of error catches and dev responses in [a PR on our fork](https://github.com/google/amphtml/pull/1).  It will get folded in to this PR when he's finished.